plugins/modules/packaging/os/pkgutil.py

```diff
@@ -59,35 +67,80 @@
 - name: Install a package from a specific repository
   community.general.pkgutil:
     name: CSWnrpe
-    site: 'ftp://myinternal.repo/opencsw/kiel'
+    site: ftp://myinternal.repo/opencsw/kiel
+    state: latest
+
+- name: Remove a package
+  community.general.pkgutil:
+    name: CSWtop
+    state: absent
+
+- name: Install several packages
+  community.general.pkgutil:
+    name:
+    - CSWsudo
+    - CSWtop
+    state: present
+
+- name: Update all packages
+  community.general.pkgutil:
+    name: '*'
+    state: latest
+
+- name: Update all packages and force versions to match latest in catalog
+  community.general.pkgutil:
+    name: '*'
     state: latest
+    force: yes
 '''
 
+RETURN = r''' # '''
+
 from ansible.module_utils.basic import AnsibleModule
 
 
-def package_installed(module, name):
-    cmd = ['pkginfo']
-    cmd.append('-q')
-    cmd.append(name)
-    rc, out, err = run_command(module, cmd)
-    if rc == 0:
-        return True
-    else:
-        return False
+def packages_not_installed(module, names):
+    ''' Check if each package is installed and return list of the ones absent '''
+    pkgs = []
+    for pkg in names:
+        rc, out, err = run_command(module, ['pkginfo', '-q', pkg])
+        if rc != 0:
+            pkgs.append(pkg)
+    return pkgs
+
 
+def packages_installed(module, names):
+    ''' Check if each package is installed and return list of the ones present '''
+    pkgs = []
+    for pkg in names:
+        if not pkg.startswith('CSW'):
+            continue
+        rc, out, err = run_command(module, ['pkginfo', '-q', pkg])
+        if rc == 0:
+            pkgs.append(pkg)
+    return pkgs
 
-def package_latest(module, name, site):
-    # Only supports one package
-    cmd = ['pkgutil', '-U', '--single', '-c']
+
+def packages_not_latest(module, names, site, update_catalog):
+    ''' Check status of each package and return list of the ones with an upgrade available '''
+    cmd = ['pkgutil']
+    if update_catalog:
+        cmd.append('-U')
```

Comment:
Should this be specified if check mode is enabled? I don't know pkgutil, but I guess this actually does modify something.

Reply:
I think it is reasonable and expected that the catalog may be updated as a part of checking for new packages.  It's just a cached copy of the list of packages available in the repository.

`-U` means "always refresh the catalog, even if it was refreshed recently".  Even without `-U`, it's quite likely that the catalog will be refreshed.

Reply:
The documentation about check mode is pretty explicit: https://docs.ansible.com/ansible/latest/user_guide/playbooks_checkmode.html#validating-tasks-check-mode-and-diff-mode

> In check mode, Ansible runs **without making any changes** on remote systems.

(emphasis mine). Modules that implement check mode have to make sure they do not modify the system when run in check mode. If I would be a user of this module, I would consider such a behavior a serious bug.

Reply:
Some writes to disk are clearly permitted, since Ansible cannot run without creating temporary files or generating logs on the remote system.

The Ansible `yum` module exhibits the same behaviour, so I think people must consider this to be acceptable.

Reply:
If you really want to do this, you should at least document this behavior.

Reply:
I’m not sure it’s helpful to obfuscate the documentation with implementation details that users aren’t going to care about or be surprised by, but okay.